src/rulesets/Base/fastmath_able.jl

```diff
@@ -163,15 +162,46 @@ let
         @scalar_rule x + y (true, true)
         @scalar_rule x - y (true, -1)
         @scalar_rule x / y (one(x) / y, -(Ω / y))
-        #log(complex(x)) is required so it gives correct complex answer for x<0
-        @scalar_rule(x ^ y,
-            (ifelse(iszero(x), zero(Ω), y * Ω / x), Ω * log(complex(x))),
-        )
-        # x^y for x < 0 errors when y is not an integer, but then derivative wrt y
-        # is undefined, so we adopt subgradient convention and set derivative to 0.
-        @scalar_rule(x::Real ^ y::Real,
-            (ifelse(iszero(x), zero(Ω), y * Ω / x), Ω * log(oftype(Ω, ifelse(x ≤ 0, one(x), x)))),
-        )
+
+        ## power
+        # literal_pow is in base.jl
+        function frule((_, Δx, Δp), ::typeof(^), x::Number, p::Number)
+            y = x ^ p
+            thegrad = _pow_grad_x(x, p, float(y))
+            thelog = if Δp isa AbstractZero
+                # Then don't waste time computing log
+                Δp
+            else
+                # When x < 0 && isinteger(p), could decide p is non-differentiable,
+                # isolated points, or could match what the rrule with ProjectTo gives:
+                _pow_grad_p(x, p, float(y))
+            end
+            return y, muladd(thelog, Δp, thegrad * Δx)
+        end
+
+        function rrule(::typeof(^), x::Number, p::Number)
+            y = x^p
+            project_x, project_p = ProjectTo(x), ProjectTo(p)
+            @inline function power_pullback(dy)
+                dx = project_x(conj(_pow_grad_x(x,p,float(y))) * dy)
+                dp = @thunk project_p(conj(_pow_grad_p(x,p,float(y))) * dy)
+                return (NoTangent(), dx, dp)
+            end
+            return y, power_pullback
+        end
+
+        _pow_grad_x(x, p, y) = (p * y / x)
+        function _pow_grad_x(x::Real, p::Real, y)
+            return ifelse(!iszero(x) | (p<0), (p * y / x),
+                     ifelse(isone(p), one(y),
+                       ifelse(0<p<1, oftype(y, Inf), zero(y) )))
+        end
+        _pow_grad_p(x, p, y) = y * log(complex(x))
+        function _pow_grad_p(x::Real, p::Real, y)
+            return ifelse(!iszero(x), y * real(log(complex(x))),
+                     ifelse(p>0, zero(y), oftype(y, NaN) ))
```

Comment:
Same here, this probably should use branching instead of `ifelse`.

Reply:
A quick attempt to time this, with contradoctory results. 
```
julia> _pow_ifelse(x::Real, p::Real, y)= ifelse(!iszero(x) | (p<0), (p * y / x),
                            ifelse(isone(p), one(y),
                              ifelse(0<p<1, oftype(y, Inf), zero(y) )));

julia> _pow_branch(x::Real, p::Real, y)= (!iszero(x) | (p<0)) ? (p * y / x) :
                            ifelse(isone(p), one(y),
                              ifelse(0<p<1, oftype(y, Inf), zero(y) ));

julia> N = 1000; x,y,p,z = randn(N),randn(N),randn(N),randn(N);

julia> _pow_ifelse.(x,p,y) ≈ _pow_branch.(x,p,y)
true

julia> @btime @. $z = _pow_ifelse($x,$p,$y);
  1.625 μs (0 allocations: 0 bytes)

julia> @btime @. $z = _pow_branch($x,$p,$y);
  653.632 ns (0 allocations: 0 bytes)

julia> x[randn(N).>0] .= 0; p[rand(N).>0.9] .= 1;

julia> @btime @. $z = _pow_branch($x,$p,$y);
  997.200 ns (0 allocations: 0 bytes)

julia> versioninfo()
Julia Version 1.7.0-beta3.0
  OS: macOS (x86_64-apple-darwin18.7.0)  # rosetta
```
Natively, the reverse? 
```
julia> @btime @. $z = _pow_ifelse($x,$p,$y);
  491.835 ns (0 allocations: 0 bytes)

julia> @btime @. $z = _pow_branch($x,$p,$y);
  634.312 ns (0 allocations: 0 bytes)

julia> x[randn(N).>0] .= 0; p[rand(N).>0.9] .= 1;

julia> @btime @. $z = _pow_branch($x,$p,$y);
  973.059 ns (0 allocations: 0 bytes)

julia> versioninfo()
Julia Version 1.8.0-DEV.459
  OS: macOS (arm64-apple-darwin20.6.0)  # M1 native
```
Remotely:
```
julia> @btime @. $z = _pow_ifelse($x,$p,$y);
  2.545 μs (0 allocations: 0 bytes)

julia> @btime @. $z = _pow_branch($x,$p,$y);
  1.177 μs (0 allocations: 0 bytes)

julia> x[randn(N).>0] .= 0; p[rand(N).>0.9] .= 1;

julia> @btime @. $z = _pow_branch($x,$p,$y);
  1.216 μs (0 allocations: 0 bytes)

julia> versioninfo()
Julia Version 1.6.1
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Gold 6226 CPU @ 2.70GHz
```
Not sure this is a good way to time this. Will a broadcast gradient ever compile down to just this or will there always be a lot of other junk around?

Reply:
Oh, very interesting! So it does look like the `ifelse` can be faster in some situations, but I think I'd still go with the branching version since it seems faster on the more common architectures and ideally llvm would figure out how to better vectorize the branching version if it can prove that there are no side effects.

Reply:
What I find most surprising is that in the first test, where the first condition is always true, the `ifelse` is path is sometimes much slower. I expected the stuff like `oftype(y, Inf), zero(y)` to be essentially free compared to `p * y / x`. 

Maybe the other tests like `isone(p)` are more expensive than I thought? In which case maybe you want more branches that just the one I tried here.

Reply:
I think this is telling me that `0<p<1` has a branch, maybe much of the above is about how that gets treated:
```
julia> Meta.@lower 1<2<3
:($(Expr(:thunk, CodeInfo(
    @ none within `top-level scope`
1 ─ %1 = 1 < 2
└──      goto #3 if not %1
2 ─ %3 = 2 < 3
└──      return %3
3 ─      return false
))))
```

Reply:
Ah, good catch! You could try it with `(0 < p) & (p < 1)` instead. These are really just microoptimizations though, so we can always leave that for later to decide.